presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -0,0 +1,421 @@
+        columnsCache = cacheBuilder.build();
+    }
+
+    @Override
+    public Set<String> getSchemaNames(JdbcIdentity identity)
+    {
+        return get(schemaNamesCache, identity);
+    }
+
+    @Override
+    public List<SchemaTableName> getTableNames(JdbcIdentity identity, Optional<String> schema)
+    {
+        return get(tableNamesCache, new TableNamesCacheKey(identity, schema));
+    }
+
+    @Override
+    public List<JdbcColumnHandle> getColumns(ConnectorSession session, JdbcTableHandle tableHandle)
+    {
+        ColumnsCacheKey key = new ColumnsCacheKey(JdbcIdentity.from(session), tableHandle);
+        List<JdbcColumnHandle> columns = columnsCache.getIfPresent(key);
+        if (columns != null) {
+            return columns;
+        }
+        columns = delegate.getColumns(session, tableHandle);
+        columnsCache.put(key, columns);
+        return columns;
+    }
+
+    @Override
+    public Optional<ColumnMapping> toPrestoType(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        return delegate.toPrestoType(session, connection, typeHandle);
+    }
+
+    @Override
+    public WriteMapping toWriteMapping(ConnectorSession session, Type type)
+    {
+        return delegate.toWriteMapping(session, type);
+    }
+
+    @Override
+    public ConnectorSplitSource getSplits(JdbcIdentity identity, JdbcTableHandle tableHandle)
+    {
+        return delegate.getSplits(identity, tableHandle);
+    }
+
+    @Override
+    public Connection getConnection(JdbcIdentity identity, JdbcSplit split)
+            throws SQLException
+    {
+        return delegate.getConnection(identity, split);
+    }
+
+    @Override
+    public void abortReadConnection(Connection connection)
+            throws SQLException
+    {
+        delegate.abortReadConnection(connection);
+    }
+
+    @Override
+    public PreparedStatement buildSql(ConnectorSession session, Connection connection, JdbcSplit split, JdbcTableHandle table, List<JdbcColumnHandle> columns)
+            throws SQLException
+    {
+        return delegate.buildSql(session, connection, split, table, columns);
+    }
+
+    @Override
+    public boolean supportsLimit()
+    {
+        return delegate.supportsLimit();
+    }
+
+    @Override
+    public boolean isLimitGuaranteed()
+    {
+        return delegate.isLimitGuaranteed();
+    }
+
+    @Override
+    public Optional<JdbcTableHandle> getTableHandle(JdbcIdentity identity, SchemaTableName schemaTableName)
+    {
+        TableHandleCacheKey key = new TableHandleCacheKey(identity, schemaTableName);
+        Optional<JdbcTableHandle> cachedTableHandle = tableHandleCache.getIfPresent(key);
+        //noinspection OptionalAssignedToNull
+        if (cachedTableHandle != null) {
+            return cachedTableHandle;
+        }
+        Optional<JdbcTableHandle> tableHandle = delegate.getTableHandle(identity, schemaTableName);
+        if (!tableHandle.isPresent() || cacheMissing) {
```

Comment:
@kokosing 
shouldn't this line be without negation?
```
if (tableHandle.isPresent() || cacheMissing) {
```
which would mean that if there is a value - cache it. if not, then check if we need to cache an empty value (cacheMissing == true), if yes - cache it.

Seems that currently it always stores empty values and only empty values.

Reply:
https://github.com/prestosql/presto/pull/6081

Reply:
It looks like you are right. Thanks!